### PR TITLE
Add client to get authenticators configured for an application

### PR DESCRIPTION
**Pull request by kayathiri4**
### Proposed changes in this pull request
$subject

**Related issue**
- https://github.com/wso2-enterprise/asgardeo-product/issues/12163

### CODE REVIEW

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.mgt.endpoint.util.client;
+
```
Comment (inthirakumaaran) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java:20]:
```suggestion
```

**Review by inthirakumaaran — APPROVED** [review]

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClient.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * Client which retrieves configure authenticators of an application.
+ */
+public class ConfiguredAuthenticatorsRetrievalClient {
+
+    private static final Log log = LogFactory.getLog(ConfiguredAuthenticatorsRetrievalClient.class);
+    private static final String APPLICATION_API_RELATIVE_PATH = "/api/server/v1/applications";
+    private static final String AUTHENTICATORS = "/authenticators";
+    private static final String SUPER_TENANT = "carbon.super";
```
Comment (AnuradhaSK) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClient.java:50]: If possible use `MultitenantConstants.SUPER_TENANT_DOMAIN_NAME`

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClient.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    public JSONObject getConfiguredAuthenticators(String applicationId)
+            throws ApplicationDataRetrievalClientException {
+
+        try (CloseableHttpClient httpclient = HttpClientBuilder.create().useSystemProperties().build()) {
+            HttpGet request =
+                    new HttpGet(getApplicationsEndpoint() + "/" + Encode.forUriComponent(applicationId) +
+                            AUTHENTICATORS);
+            setAuthorizationHeader(request);
+
+            try (CloseableHttpResponse response = httpclient.execute(request)) {
```
Comment (AnuradhaSK) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClient.java:69]: Can't we add this resource into the above try with resource block ?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Exception thrown on a client error when retrieving authenticators configured for an application.
+ */
+public class ConfiguredAuthenticatorsRetrievalClientException extends Exception {
+
+    public ConfiguredAuthenticatorsRetrievalClientException(String message) {
+        super(message);
```
Comment (AnuradhaSK) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java:28]: missing new line after method signature. fix in other places as wel

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.mgt.endpoint.util.client;
+
+
+/**
+ * Exception thrown on a client error when retrieving authenticators configured for an application.
+ */
+public class ConfiguredAuthenticatorsRetrievalClientException extends Exception {
```
Comment (AnuradhaSK) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java:25]: Don't we support error codes?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java (hunk 1 above)
Comment (kayathiri4) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java:20]: Do we need to remove this?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java (hunk 5 above)
Comment (kayathiri4) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClientException.java:25]: We do not use them specifically

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClient.java (hunk 3 above)
Comment (kayathiri4) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/ConfiguredAuthenticatorsRetrievalClient.java:70]: We need to do some more actions in between both blocks

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**